File: ZKCache/src/main/java/com/ericsson/ema/tim/dml/predicate/UnEq.java
package com.ericsson.ema.tim.dml.predicate;

import com.ericsson.ema.tim.dml.DataTypes;
import com.ericsson.ema.tim.exception.DmlBadSyntaxException;
import com.ericsson.ema.tim.exception.DmlNoSuchFieldException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

/**
 * Created by eqinson on 2017/4/23.
 */
public class UnEq extends AbstractPredicate implements Predicate {
    private final static Logger LOGGER = LoggerFactory.getLogger(UnEq.class);

    private UnEq(String field, String value) {
        super(field, value);
    }

    public static UnEq uneq(String field, String value) {
        return new UnEq(field, value);
    }

    @Override
    public boolean eval(Object tuple) {
        if (this.valueToComp == null)
            return true;

        Object fieldVal = getFiledValFromTupleByName(tuple);
        Map<String, String> metadata = getSelector().getContext().getTableMetadata();
        String fieldType = metadata.get(field);
        if (fieldType == null)
            throw new DmlNoSuchFieldException(field);

        switch (fieldType) {
            case DataTypes.String:
                return !(this.valueToComp).equals(fieldVal);
            case DataTypes.Int:
                return !(Integer.valueOf((String) this.valueToComp)).equals(fieldVal);
            case DataTypes.Boolean:
                return !(Boolean.valueOf((String) this.valueToComp)).equals(fieldVal);
            default:
                LOGGER.error("unsupported data type: {},{}", field, fieldType);
                throw new DmlBadSyntaxException("unsupported data type: " + field + "," + fieldType);
        }
    }
}

File: zkcache_scala/src/main/scala/com/ericsson/ema/tim/dml/predicate/UnEq.scala
package com.ericsson.ema.tim.dml.predicate

import com.ericsson.ema.tim.dml.DataTypes
import com.ericsson.ema.tim.exception.{DmlBadSyntaxException, DmlNoSuchFieldException}
import org.slf4j.LoggerFactory

/**
  * Created by eqinson on 2017/5/13.
  */
class UnEq private(override val field: String, override val valueToComp: Object) extends AbstractPredicate(field, valueToComp) with Predicate {
    private val LOGGER = LoggerFactory.getLogger(classOf[UnEq])

    override def eval(tuple: Object): Boolean = {
        if (this.valueToComp eq null)
            return true

        val fieldVal: Object = getFiledValFromTupleByName(tuple)
        val fieldType = selector.context.tableMetadata.get(field)
        fieldType match {
            case Some(DataTypes.String)  =>
                this.valueToComp != fieldVal
            case Some(DataTypes.Int)     =>
                Integer.valueOf(this.valueToComp.asInstanceOf[String]) != fieldVal.asInstanceOf[Integer]
            case Some(DataTypes.Boolean) =>
                java.lang.Boolean.valueOf(this.valueToComp.asInstanceOf[String]) != fieldVal.asInstanceOf[java.lang.Boolean]
            case Some(_)                 =>
                LOGGER.error("unsupported data type: {},{}", field, fieldType: Any)
                throw DmlBadSyntaxException("unsupported data type: " + field + "," + fieldType)
            case None                    => throw DmlNoSuchFieldException(field)
        }
    }
}

object UnEq {
    def apply(field: String, value: String) = new UnEq(field, value)
}